sdk/storage/Azure.Storage.Blobs/api/Azure.Storage.Blobs.netstandard2.0.cs

```diff
@@ -976,6 +976,8 @@ public partial class BlobBaseClient
         public virtual System.Threading.Tasks.Task<Azure.Response<bool>> ExistsAsync(System.Threading.CancellationToken cancellationToken = default(System.Threading.CancellationToken)) { throw null; }
         public virtual Azure.Response<Azure.Storage.Blobs.Models.BlobProperties> GetProperties(Azure.Storage.Blobs.Models.BlobRequestConditions conditions = null, System.Threading.CancellationToken cancellationToken = default(System.Threading.CancellationToken)) { throw null; }
         public virtual System.Threading.Tasks.Task<Azure.Response<Azure.Storage.Blobs.Models.BlobProperties>> GetPropertiesAsync(Azure.Storage.Blobs.Models.BlobRequestConditions conditions = null, System.Threading.CancellationToken cancellationToken = default(System.Threading.CancellationToken)) { throw null; }
+        public virtual System.IO.Stream OpenRead(long position = (long)0, int bufferSize = 1048576, Azure.Storage.Blobs.Models.BlobRequestConditions conditions = null) { throw null; }
```

Comment:
How is this different from Download/DownloadAsync?

Reply:
`OpenRead()` returns a stream that will download the blob/file as it's `Read()` method is called.  `Download()`/`DownloadAsync()` download the entire blob/file (or a range of it) all at once.

Let's say I have a 100 GB blob, and I'd like to download it, and then preform several other operations on it, but only have 16 GB of ram.  I can do so using `OpenRead()`, and don't have to manually manage the ranged download calls myself.

The lack of this functionality is blocking a lot of customers from upgrading to V12.

See the discussion here - https://github.com/Azure/azure-sdk-for-net/issues/9607

Reply:
> Download()/DownloadAsync() download the entire blob/file (or a range of it) all at once.

Are you sure that is true? We don't read data off the network connection until read is called on the stream itself.

Reply:
And if that's true then `Download[Async]` needs fixing.

Reply:
https://github.com/Azure/azure-sdk-for-net/blob/master/sdk/storage/Azure.Storage.Blobs/tests/BlobBaseClientTests.cs#L149

Pretty sure.  Change `var data = GetRandomBuffer(Constants.KB);` to `var data = GetRandomBuffer(Constants.GB);`, and run the test with Fiddler.  A massive get request is sent on line 162, before we read for the resulting stream.

Reply:
Get request can not be massive because it doesn't have a body. 

>In addition, it's more optimal for us to send multiple, smaller gets. 

Why?

Reply:
Massive response.  Got me on the typo.

We've found its faster to download large blobs in multiple parts than in one large request.

@pakrym, please read the discussion on the GitHub issue - https://github.com/Azure/azure-sdk-for-net/issues/9607.

Reply:
So I wrote this sample app that shows that process memory doesn't increase when reading the large blob:

```
using System;
using Azure.Storage.Blobs;
using System.Threading.Tasks;
using System.Diagnostics;

namespace azurelogs
{
    class Program
    {
        static async Task Main(string[] args)
        {
            var value = await new BlobClient(new Uri("https://pakrym0test0storage.blob.core.windows.net/myblobcontainer/VS2015 Image Library.zip")).DownloadAsync();
            var stream = value.Value.Content;

            var buffer = new byte[1024 * 1024];
            int read;
            int total = 0;
            var process = Process.GetCurrentProcess();
            while ((read = await stream.ReadAsync(buffer)) > 0)
            {
                total += read;
                Console.WriteLine("Downloaded "+ total / 1024 / 1024 + "MB, Process memory " + process.PrivateMemorySize64 / 1024 / 1024 + "MB");
            }
        }
    }
}

```

Feel free to run it, it doesn't require auth.

Reply:
@pakrym The edge `OpenRead` has is that it frees up connection between reads. Let's say I'm merging 1000 files by polling one line from each at a time (merge N sorted lists let's say) then I'd need to open 1000 connections to storage in order to get this going (otherwise my requets are going to queue since some streams are not yet exhaousted) if I only used `Download`.

Now how do we want to present this feature. I think it could be either extra parameter passed to `Download` where we say `lazyStream` or something OR it could be separate API like in this PR. If that feature was/is popular in track 1 then the later is more discover-able IMO.

Reply:
>The edge OpenRead has is that it frees up connection between reads

I don't see the current implementation doing what you are saying, it seems to start a download and then hold onto the network stream in between reads.

In your scenario would you prefer you connection reestablished 1000x(line count) times instead? 

In general it would be nice to know what problem is being solved here, I've seen multiple being mentioned:
1. Memory usage
2. Requesting smaller blocks vs entire blob
3. Returning connection in between reads.

Reply:
Good point. It seems that `LazyLoadingReadOnlyStream` does hold onto network stream. I the track one network stream is immediately read into memory buffer ( see https://github.com/Azure/azure-storage-net/blob/319b1a03e6ec2304a0a80caf226ba385f8ba1150/Lib/ClassLibraryCommon/Blob/BlobReadStream.cs#L141 ). @seanmcc-msft I think we should take a look on that aspect and fix it.

The intention isn't to reestablish 1000 connections but rather not clog them. I.e. un-disposed network stream is going to make connection not available for other thread on .NET Framework.
In that hypothetical scenario. If processing a line takes long before requesting next one (line processing time >>> request line) then we could process that workload with opening less connections but more request-response roundtrips.

Reply:
Can you link to track 1 implementation of this feature please.

Reply:
- https://github.com/Azure/azure-storage-net/blob/master/Lib/ClassLibraryCommon/Blob/CloudBlob.cs#L49
- https://github.com/Azure/azure-storage-net/blob/master/Lib/ClassLibraryCommon/Blob/BlobReadStream.cs
- https://github.com/Azure/azure-storage-net/blob/master/Lib/Common/Blob/BlobReadStreamBase.cs
- https://github.com/Azure/azure-storage-net/blob/master/Lib/Common/Core/MultiBufferMemoryStream.cs

Reply:
[The API](https://github.com/Azure/azure-storage-net/blob/319b1a03e6ec2304a0a80caf226ba385f8ba1150/Lib/ClassLibraryCommon/Blob/CloudBlob.cs#L49-L54)
[ReadStream](https://github.com/Azure/azure-storage-net/blob/319b1a03e6ec2304a0a80caf226ba385f8ba1150/Lib/ClassLibraryCommon/Blob/BlobReadStream.cs#L141)